From 38b823814c9c7a2cbccb7dd520796729ebd11d8c Mon Sep 17 00:00:00 2001
From: Dmitry Torokhov <dtor@chromium.org>
Date: Thu, 5 Jan 2017 14:48:47 -0800
Subject: [PATCH 32/35] CHROMIUM: sched: tune: allow nested schedtune groups

Let's remove restriction that schedtune groups can not be nested. Note
that they are still completely independent even if they are nested, but
at least we can build proper hierarchies now.

We also extend maximum number of schedtune groups to 16 for now.

CHROMIUM Note: Apparently the plan is to drop schedtune in favor of
utilization clamping. Future kernel rebases should hopefully be able
to drop this patch once utilization clamping is accepted upstream and
we've moved to it. Check out the second bug in the BUG= field.

BUG=b:74017669 b:116869240
TEST=Build and boot Cheza with EAS enabled

Signed-off-by: Dmitry Torokhov <dtor@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/425443
Reviewed-by: Sonny Rao <sonnyrao@chromium.org>
Change-Id: Icdf8db99dc2d252b782b228caedefa6abf41a9a6
Reviewed-on: https://chromium-review.googlesource.com/1235093
Commit-Ready: Evan Green <evgreen@chromium.org>
Tested-by: Evan Green <evgreen@chromium.org>
Reviewed-by: Douglas Anderson <dianders@chromium.org>
---
 kernel/sched/tune.c | 8 +-------
 1 file changed, 1 insertion(+), 7 deletions(-)

diff --git a/kernel/sched/tune.c b/kernel/sched/tune.c
index 5670542c774e..2eae0813c034 100644
--- a/kernel/sched/tune.c
+++ b/kernel/sched/tune.c
@@ -105,7 +105,7 @@ root_schedtune = {
  *    implementation especially for the computation of the per-CPU boost
  *    value
  */
-#define BOOSTGROUPS_COUNT 6
+#define BOOSTGROUPS_COUNT 16
 
 /* Array of configured boostgroups */
 static struct schedtune *allocated_group[BOOSTGROUPS_COUNT] = {
@@ -708,12 +708,6 @@ schedtune_css_alloc(struct cgroup_subsys_state *parent_css)
 	if (!parent_css)
 		return &root_schedtune.css;
 
-	/* Allow only single level hierachies */
-	if (parent_css != &root_schedtune.css) {
-		pr_err("Nested SchedTune boosting groups not allowed\n");
-		return ERR_PTR(-ENOMEM);
-	}
-
 	/* Allow only a limited number of boosting groups */
 	for (idx = 1; idx < BOOSTGROUPS_COUNT; ++idx)
 		if (!allocated_group[idx])
-- 
2.39.2

